backend/v3/domain/settings.go

```diff
@@ -0,0 +1,286 @@
+	MaxOTPAttempts      uint64 `json:"maxOtpAttempts,omitempty"`
+	ShowLockOutFailures bool   `json:"showLockOutFailures,omitempty"`
+}
+
+type LockoutSetting struct {
+	*Setting
+	Settings LockoutSettings
+}
+
+type DomainSettings struct {
+	IsDefault                              bool `json:"isDefault,omitempty"`
+	UserLoginMustBeDomain                  bool `json:"userLoginMustBeDomain,omitempty"`
+	ValidateOrgDomains                     bool `json:"validateOrgDomains,omitempty"`
+	SMTPSenderAddressMatchesInstanceDomain bool `json:"smtpSenderAddressMatchesInstanceDomain,omitempty"`
+}
+
+type DomainSetting struct {
+	*Setting
+	Settings DomainSettings
+}
+
+type SecuritySettings struct {
+	Enabled               bool     `json:"enabled,omitempty"`
+	EnableIframeEmbedding bool     `json:"enableIframe_embedding,omitempty"`
+	AllowedOrigins        []string `json:"allowedOrigins,omitempty"`
+	EnableImpersonation   bool     `json:"enableImpersonation,omitempty"`
+}
+
+type SecuritySetting struct {
+	*Setting
+	Settings SecuritySettings
+}
+
+type OrgSettings struct {
+	OrganizationScopedUsernames    bool     `json:"organizationScopedUsernames,omitempty"`
+	OldOrganizationScopedUsernames bool     `json:"oldOrganizationScopedUsernames,omitempty"`
+	UsernameChanges                []string `json:"usernameChanges,omitempty"`
+}
+
+type OrgSetting struct {
+	*Setting
+	Settings OrgSettings
+}
+
+type settingsColumns interface {
+	IDColumn() database.Column
+	InstanceIDColumn() database.Column
+	OrgIDColumn() database.Column
+	TypeColumn() database.Column
+	LabelStateColumn() database.Column
+	SettingsColumn() database.Column
+	CreatedAtColumn() database.Column
+	UpdatedAtColumn() database.Column
+}
+
+type settingsConditions interface {
+	InstanceIDCondition(id string) database.Condition
+	OrgIDCondition(id *string) database.Condition
+	IDCondition(id string) database.Condition
+	TypeCondition(typ SettingType) database.Condition
+	LabelStateCondition(typ LabelState) database.Condition
+}
+
+type Settings interface {
+	GetSettings() []byte
+}
+
+type settingsChanges interface {
+	SetType(state SettingType) database.Change
+	SetSettings(settings string) database.Change
+	SetUpdatedAt(updatedAt *time.Time) database.Change
+}
+
+type SettingsRepository interface {
+	settingsColumns
+	settingsConditions
+	settingsChanges
+
+	Get(ctx context.Context, client database.QueryExecutor, instanceID string, orgID *string, typ SettingType, cond ...database.Condition) (*Setting, error)
+	List(ctx context.Context, client database.QueryExecutor, conditions ...database.Condition) ([]*Setting, error)
+
+	CreateLogin(ctx context.Context, client database.QueryExecutor, setting *LoginSetting) error
+	GetLogin(ctx context.Context, client database.QueryExecutor, instanceID string, orgID *string) (*LoginSetting, error)
+	UpdateLogin(ctx context.Context, client database.QueryExecutor, setting *LoginSetting, changes ...database.Change) (int64, error)
```

Comment:
Preferable we should only have a `Set` method otherwise the business logic must do multiple calls (check if a policy exists and then update or create it)
```suggestion
	SetLogin(ctx context.Context, client database.QueryExecutor, setting *LoginSetting) error
	GetLogin(ctx context.Context, client database.QueryExecutor, instanceID string, orgID *string) (*LoginSetting, error)
```

Reply:
as we anyways need to touch the repo i would split it into the different settings types. e.g.

```go
type LoginSettingsRepo interface {
    Get(...)
    Set(...)
}

type LabelSettingsRepo interface {
    Get(...)
    Set(...)
    Activate(...)
}
```

Reply:
I guess we can still have a single method for `Get` and `Reset` in the repository package by just adding a type to the condition before calling the private function.